contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,158 @@
+
+## `Collection` Resource
+
+The resource that stores a user's NFT collection.
+It includes a few functions to allow the owner to easily
+move tokens in and out of the collection.
+
+## `Provider` and `Receiver` resource interfaces
+
+These interfaces declare functions with some pre and post conditions
+that require the Collection to follow certain naming and behavior standards.
+
+They are separate because it gives the user the ability to share a reference
+to their Collection that only exposes the fields and functions in one or more
+of the interfaces. It also gives users the ability to make custom resources
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+pub contract NonFungibleToken {
+
+    /// Interface that the NFTs have to conform to
+    ///
+    pub resource interface NFT { //: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub let id: UInt64
+
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
+        }
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{Receiver}>): @AnyResource{NFT}
+    }
+
+    /// Interface to mediate deposits to the Collection
+    ///
+    pub resource interface Receiver {
+
+        /// deposit takes an NFT as an argument and adds it to the Collection
+        ///
+        pub fun deposit(token: @AnyResource{NFT})
+
+        /// getAcceptedTypes returns a list of NFT types that this receiver accepts
+        pub fun getAcceptedTypes(): [Type]
+    }
+
+    /// Interface that an account would commonly 
+    /// publish for their collection
+    pub resource interface CollectionPublic { //: MetadataViews.ResolverCollection {
+        pub fun deposit(token: @AnyResource{NFT})
+        pub fun borrowViewResolver(id: UInt64): &{MetadataViews.Resolver}
```

Comment:
this can be removed probably

Reply:
Why?

Reply:
I'd guess we won't need this definition if the inheritance that's commented out is taken? Or are we deprecating that interface altogether?

Reply:
it is only commented out right now because cadence currently doesn't support interfaces inheriting from other interfaces.

Reply:
yes, the cadence team is working on it

Reply:
oh I didn't see this before sorry, I was telling to remove this totally I remember ( not related to interface ) , but now I have to dig deep in memory banks.

Reply:
I think I was thinking because now we have:
https://github.com/onflow/flow-nft/blob/d39375c13fdaf23977d813d9a357c4afe08c8125/contracts/NonFungibleToken-v2.cdc#L56-57

result of borrowNFT can resolve views, we don't need resolver anymore . It was needed because we couldn't downcast ( or upcast dont know the terminology ) NFT to MetadataResolver

Reply:
really? I thought that the view resolver for the collection was still nice to have

Reply:
I maybe missing the context, @bjartek knows better, but I think it was like that.

Reply:
I think keeping the same name as v1 has merrit here, since it would make it easier to reason about code using both v1 and v2 of the spec. When that is said that might not be a goal that we can achieve. 

I really want to try out different alternatives when we have default implementations and see.

Reply:
> I think keeping the same name as v1 has merrit here, since it would make it easier to reason about code using both v1 and v2 of the spec. When that is said that might not be a goal that we can achieve.

But to be clear, you're talking about maintaining consistency between the V1 and V2 specs, right? I'm not sure this is important if we already have to upgrade all NFT implementations from V1 to V2.

Apart from that, does the `MetadataViews.ResolverCollection` interface continue to have value on its own?

IMO it doesn't have any real value now that the NFT interface implements `ViewResolver`. I think we should remove the `borrowViewResolver` function to keep the collection interface as small as possible.